File: algorithm/src/polynomial.py
def naive(coefficients, x):
    res = 0
    # a_0 + a_1x + a_2x^2 + ... + a_{n-1}x^{n-1} + a_nx^n
    for exp, coefficient in enumerate(coefficients[1:], start=0):
        res += pow(x, exp) * coefficient
    return res


def better(coefficients, x):
    res = 0
    # a_0 + x(a_1 + x(...(a_{n-1} + x(a_n))))
    for coefficient in coefficients[::-1]:
        res += x * (coefficient + res)
    return res

File: algorithm/src/Polynomial.c
#include <stdio.h>
#include <time.h>
#include <math.h>

clock_t start, stop;
double duration;
#define MAXN 10;


double f(double a[], double x){
    // a is coefficients
    // x is value
    double res = 0.0;
    int n = sizeof a;
    for(int i = 0;i <= n;i++){
        res += (a[i] * pow(x, i));
    }
    return res;
}

double f2(double a[], double x){
    double res = 0.0;
    int n = sizeof a;
    for (int i = n;i >= 0; i--) {
        res += x * (a[i] + res);
    }
    return res;
}